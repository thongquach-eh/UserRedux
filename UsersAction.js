//@flow
import type {UserAction} from './types.js';

export const addUser = (): UserAction => ({
  type: 'ADD_USER',
});
