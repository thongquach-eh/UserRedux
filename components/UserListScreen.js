// @flow
import * as React from 'react';
import {FlatList} from 'react-native';
import type {RouteProp} from '@react-navigation/native';
import type {StackNavigationProp} from '@react-navigation/stack';
import type {UserState} from '../types.js';
import {useSelector} from 'react-redux';
import UserItem from './UserItem';

type UserListScreenProps = {
  navigation: StackNavigationProp<any, any>,
  route: RouteProp<any, any>,
};

const UserListScreen = ({
  navigation,
  route,
}: UserListScreenProps): React.Node => {
  const users = useSelector((state: UserState) => state.users);

  return (
    <FlatList
      data={users}
      renderItem={({item: user}) => (
        <UserItem user={user} navigation={navigation} />
      )}
      keyExtractor={user => user.email}
    />
  );
};

export default UserListScreen;
