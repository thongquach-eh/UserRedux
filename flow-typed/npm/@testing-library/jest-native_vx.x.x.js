// flow-typed signature: 9bf09a1bcb3c6ad49aff54424c32ba18
// flow-typed version: <<STUB>>/@testing-library/jest-native_v^4.0.1/flow_v0.137.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@testing-library/jest-native'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@testing-library/jest-native' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@testing-library/jest-native/dist/extend-expect' {
  declare module.exports: any;
}

declare module '@testing-library/jest-native/dist' {
  declare module.exports: any;
}

declare module '@testing-library/jest-native/dist/to-be-disabled' {
  declare module.exports: any;
}

declare module '@testing-library/jest-native/dist/to-be-empty' {
  declare module.exports: any;
}

declare module '@testing-library/jest-native/dist/to-contain-element' {
  declare module.exports: any;
}

declare module '@testing-library/jest-native/dist/to-have-prop' {
  declare module.exports: any;
}

declare module '@testing-library/jest-native/dist/to-have-style' {
  declare module.exports: any;
}

declare module '@testing-library/jest-native/dist/to-have-text-content' {
  declare module.exports: any;
}

declare module '@testing-library/jest-native/dist/utils' {
  declare module.exports: any;
}

declare module '@testing-library/jest-native/extend-expect' {
  declare module.exports: any;
}

// Filename aliases
declare module '@testing-library/jest-native/dist/extend-expect.js' {
  declare module.exports: $Exports<'@testing-library/jest-native/dist/extend-expect'>;
}
declare module '@testing-library/jest-native/dist/index' {
  declare module.exports: $Exports<'@testing-library/jest-native/dist'>;
}
declare module '@testing-library/jest-native/dist/index.js' {
  declare module.exports: $Exports<'@testing-library/jest-native/dist'>;
}
declare module '@testing-library/jest-native/dist/to-be-disabled.js' {
  declare module.exports: $Exports<'@testing-library/jest-native/dist/to-be-disabled'>;
}
declare module '@testing-library/jest-native/dist/to-be-empty.js' {
  declare module.exports: $Exports<'@testing-library/jest-native/dist/to-be-empty'>;
}
declare module '@testing-library/jest-native/dist/to-contain-element.js' {
  declare module.exports: $Exports<'@testing-library/jest-native/dist/to-contain-element'>;
}
declare module '@testing-library/jest-native/dist/to-have-prop.js' {
  declare module.exports: $Exports<'@testing-library/jest-native/dist/to-have-prop'>;
}
declare module '@testing-library/jest-native/dist/to-have-style.js' {
  declare module.exports: $Exports<'@testing-library/jest-native/dist/to-have-style'>;
}
declare module '@testing-library/jest-native/dist/to-have-text-content.js' {
  declare module.exports: $Exports<'@testing-library/jest-native/dist/to-have-text-content'>;
}
declare module '@testing-library/jest-native/dist/utils.js' {
  declare module.exports: $Exports<'@testing-library/jest-native/dist/utils'>;
}
declare module '@testing-library/jest-native/extend-expect.js' {
  declare module.exports: $Exports<'@testing-library/jest-native/extend-expect'>;
}
