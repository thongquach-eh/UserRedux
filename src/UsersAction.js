//@flow
import type {UserAction, User} from './types.js';

export const pressAddUser = (): UserAction => ({
  type: 'PRESS_ADD_USER',
});

export const addUser = (user: User): UserAction => ({
  type: 'ADD_USER',
  newUser: user,
});

export const addUsers = (users: User[]): UserAction => ({
  type: 'ADD_USERS',
  newUsers: users,
});

export const pressEditUser = (): UserAction => ({
  type: 'PRESS_EDIT_USER',
});

export const editUser = (user: User): UserAction => ({
  type: 'EDIT_USER',
  editedUser: user,
});
