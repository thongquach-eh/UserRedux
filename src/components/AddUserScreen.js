// @flow
import React, {useLayoutEffect, useCallback} from 'react';
import type {Node} from 'react';
import {
  Button,
  StyleSheet,
  ScrollView,
  KeyboardAvoidingView,
  Platform,
} from 'react-native';
import type {RouteProp} from '@react-navigation/native';
import type {StackNavigationProp} from '@react-navigation/stack';
import {useHeaderHeight} from '@react-navigation/stack';
import {useDispatch} from 'react-redux';
import {pressAddUser} from '../UsersAction';
import type {UserDispatch, RootStackParamList} from '../types.js';
import StringInput from './StringInput';
import 'react-native-get-random-values';
import {validateUser} from './Validations';
import GenderPicker from './GenderPicker';
import BirthDatePicker from './BirthDatePicker';
import LocationInput from './LocationInput';
import FullNameInput from './FullNameInput';
import PictureInput from './PictureInput';
import {reduxForm, Field} from 'redux-form';
import type {FormProps} from 'redux-form';

type AddUserScreenProps = {
  navigation: StackNavigationProp<RootStackParamList, 'AddUser'>,
  route: RouteProp<RootStackParamList, 'AddUser'>,
} & FormProps;

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  detailsContainer: {
    margin: 10,
  },
});

const AddUserScreen = (props: AddUserScreenProps): Node => {
  const {navigation, valid} = props;
  const dispatch = useDispatch<UserDispatch>();

  const saveUser = useCallback(() => {
    if (valid) {
      dispatch(pressAddUser());
      return true;
    } else {
      // eslint-disable-next-line no-alert
      alert('The form is invalid. Please recheck the entered data.');
    }
  }, [dispatch, valid]);

  useLayoutEffect(() => {
    navigation.setOptions({
      headerRight: () => (
        <Button
          title="Save"
          onPress={() => {
            if (saveUser()) {
              navigation.goBack();
            }
          }}
        />
      ),
    });
  }, [saveUser, navigation]);

  return (
    <KeyboardAvoidingView
      behavior={Platform.OS === 'ios' ? 'padding' : null}
      style={styles.container}
      keyboardVerticalOffset={useHeaderHeight() + 20}>
      <ScrollView style={styles.detailsContainer}>
        <Field
          name="picture"
          component={PictureInput}
          props={{label: 'Choose picture'}}
        />
        <FullNameInput />
        <Field
          name="gender"
          component={GenderPicker}
          props={{label: 'Gender:'}}
        />
        <BirthDatePicker label="Birth Date:" />
        <Field name="email" component={StringInput} props={{label: 'Email:'}} />
        <Field name="phone" component={StringInput} props={{label: 'Phone:'}} />
        <Field
          name="cell"
          component={StringInput}
          props={{label: 'Cellphone:'}}
        />
        <LocationInput label="Location:" />
      </ScrollView>
    </KeyboardAvoidingView>
  );
};

export default reduxForm({form: 'addUser', validate: validateUser})(
  AddUserScreen,
);
