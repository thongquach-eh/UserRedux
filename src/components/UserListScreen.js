// @flow
import * as React from 'react';
import {
  FlatList,
  Button,
  ActivityIndicator,
  View,
  StyleSheet,
  Text,
} from 'react-native';
import type {RouteProp} from '@react-navigation/native';
import type {StackNavigationProp} from '@react-navigation/stack';
import type {RootState, RootStackParamList, UserDispatch} from '../types';
import {useSelector, useDispatch} from 'react-redux';
import UserItem from './UserItem';
import {isFetchingUsersSelector, fetchUsersErrorSelector} from '../state';
import {startFetchUsers} from '../UsersAction';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
  },
  errorText: {
    textAlign: 'center',
    fontSize: 16,
  },
});

type UserListScreenProps = {
  navigation: StackNavigationProp<RootStackParamList, 'UserList'>,
  route: RouteProp<RootStackParamList, 'UserList'>,
};

const UserListScreen = ({
  navigation,
  route,
}: UserListScreenProps): React.Node => {
  const users = useSelector((state: RootState) => state.user.users);
  const dispatch = useDispatch<UserDispatch>();
  const isFetching = useSelector<RootState, boolean>(isFetchingUsersSelector);
  const fetchUsersError = useSelector<RootState, any>(fetchUsersErrorSelector);

  React.useEffect(() => {
    dispatch(startFetchUsers());
  }, [dispatch]);

  React.useLayoutEffect(() => {
    navigation.setOptions({
      headerRight: () => (
        <Button title="Add" onPress={() => navigation.navigate('AddUser')} />
      ),
    });
  });

  if (isFetching) {
    return (
      <View style={styles.container}>
        <ActivityIndicator size="large" />
      </View>
    );
  } else if (fetchUsersError) {
    return (
      <View style={styles.container}>
        <Text style={styles.errorText}>
          An error occurred: {fetchUsersError.message}
        </Text>
        <Button onPress={() => dispatch(startFetchUsers())} title="Retry" />
      </View>
    );
  } else {
    return (
      <FlatList
        data={users}
        renderItem={({item: user}) => (
          <UserItem user={user} navigation={navigation} />
        )}
        keyExtractor={user => user.id}
      />
    );
  }
};

export default UserListScreen;
