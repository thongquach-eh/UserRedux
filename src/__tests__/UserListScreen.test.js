import 'react-native';
import * as React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import UserListScreen from '../components/UserListScreen';
import AddUserScreen from '../components/AddUserScreen';
import {Provider} from 'react-redux';
import {createStore, combineReducers, applyMiddleware} from 'redux';
import usersReducer from '../UsersReducer';
import {
  reducers as apiReducers,
  middleware as apiMiddleware,
} from 'redux-api-call';
import {reducer as formReducer} from 'redux-form';
import {startFetchUsers} from '../UsersAction';
import {NavigationContainer} from '@react-navigation/native';
import {createStackNavigator} from '@react-navigation/stack';
import DateTimePicker from '@react-native-community/datetimepicker';
jest.mock('@react-native-community/datetimepicker');

const mockedNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  return {
    ...jest.requireActual('@react-navigation/native'),
    useNavigation: () => {
      const navigation = jest
        .requireActual('@react-navigation/native')
        .useNavigation();
      return {
        ...navigation,
        navigate: mockedNavigate,
      };
    },
  };
});
const configureStore = initialState => {
  const middlewares = [apiMiddleware];
  const rootReducer = combineReducers({
    user: usersReducer,
    form: formReducer,
    ...apiReducers,
  });

  const store = createStore(
    rootReducer,
    initialState,
    applyMiddleware(...middlewares),
  );

  let actions = [];
  const saveAction = next => action => {
    actions.push(action);
  };

  return {
    ...store,
    dispatch: saveAction(store.dispatch),
    getActions: () => actions,
  };
};

const renderComponent = initialState => {
  const Stack = createStackNavigator();

  const store = configureStore(initialState);
  const component = (
    <Provider store={store}>
      <NavigationContainer>
        <Stack.Navigator>
          <Stack.Screen
            name="UserList"
            component={UserListScreen}
            options={{title: 'Users'}}
          />
        </Stack.Navigator>
      </NavigationContainer>
    </Provider>
  );
  const wrapper = render(component);

  return {wrapper, store};
};

describe('User List Screen', () => {
  it('should show users', () => {
    const initialState = {
      user: {
        users: [
          {
            id: '12341214',
            gender: 'male',
            name: {
              title: 'Mr',
              first: 'Emre',
              last: 'Abacı',
            },
            location: {
              street: {
                number: 7028,
                name: 'Vatan Cd',
              },
              city: 'Hatay',
              state: 'Elazığ',
              country: 'Turkey',
              postcode: 34478,
            },
            email: 'emre.abaci@example.com',
            dob: {
              date: '1947-06-30T21:13:54.078Z',
              age: 74,
            },
            phone: '(439)-741-1246',
            cell: '(626)-097-5896',
            picture: {
              large: 'https://randomuser.me/api/portraits/men/39.jpg',
              medium: 'https://randomuser.me/api/portraits/med/men/39.jpg',
              thumbnail: 'https://randomuser.me/api/portraits/thumb/men/39.jpg',
            },
          },
        ],
      },
      api_calls: {
        FETCH_USERS: {
          isFetching: false,
          error: null,
        },
      },
    };
    const {wrapper} = renderComponent(initialState);

    expect(wrapper.queryAllByText('Mr. Emre Abacı')).toHaveLength(1);
    expect(
      wrapper.queryAllByText('Vatan Cd 7028, Hatay Elazığ, 34478 Turkey'),
    ).toHaveLength(1);
    expect(wrapper.queryAllByText('July 1, 1947')).toHaveLength(1);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('is getting users', () => {
    const initialState = {
      api_calls: {
        FETCH_USERS: {
          isFetching: true,
          error: null,
        },
      },
    };
    const {wrapper} = renderComponent(initialState);

    expect(wrapper.queryAllByTestId('loadingIcon')).toHaveLength(1);
  });

  it('get errors', () => {
    const initialState = {
      api_calls: {
        FETCH_USERS: {
          isFetching: false,
          error: {
            message: 'Something went wrong!',
          },
        },
      },
    };
    const {wrapper} = renderComponent(initialState);

    expect(
      wrapper.queryAllByText('An error occurred: Something went wrong!'),
    ).toHaveLength(1);
  });

  it('should fetch users when entering', () => {
    const {store} = renderComponent({});
    expect(store.getActions()).toContainEqual(startFetchUsers());
  });

  it('should navigate to Add User Screen 2', () => {
    const Stack = createStackNavigator();
    const store = configureStore({});
    const component = (
      <Provider store={store}>
        <NavigationContainer>
          <Stack.Navigator>
            <Stack.Screen
              name="UserList"
              component={UserListScreen}
              options={{title: 'Users'}}
            />
          </Stack.Navigator>
        </NavigationContainer>
      </Provider>
    );
    const wrapper = render(component);

    const addUserBtn = wrapper.getByText('Add'); //error happened here because Add button does not exist without Navigation Container
    fireEvent.press(addUserBtn);

    expect(mockedNavigate).toHaveBeenCalledWith('AddUser');
  });
});
