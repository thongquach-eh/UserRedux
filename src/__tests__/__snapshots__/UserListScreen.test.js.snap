// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User List Screen should show users 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "cell": "(626)-097-5896",
        "dob": Object {
          "age": 74,
          "date": "1947-06-30T21:13:54.078Z",
        },
        "email": "emre.abaci@example.com",
        "gender": "male",
        "id": "12341214",
        "location": Object {
          "city": "Hatay",
          "country": "Turkey",
          "postcode": 34478,
          "state": "Elazığ",
          "street": Object {
            "name": "Vatan Cd",
            "number": 7028,
          },
        },
        "name": Object {
          "first": "Emre",
          "last": "Abacı",
          "title": "Mr",
        },
        "phone": "(439)-741-1246",
        "picture": Object {
          "large": "https://randomuser.me/api/portraits/men/39.jpg",
          "medium": "https://randomuser.me/api/portraits/med/men/39.jpg",
          "thumbnail": "https://randomuser.me/api/portraits/thumb/men/39.jpg",
        },
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <View
          style={
            Object {
              "backgroundColor": "lightblue",
              "flexDirection": "row",
              "marginHorizontal": 8,
              "marginVertical": 8,
              "padding": 12,
            }
          }
        >
          <Image
            source={
              Object {
                "uri": "https://randomuser.me/api/portraits/med/men/39.jpg",
              }
            }
            style={
              Object {
                "height": 72,
                "width": 72,
              }
            }
          />
          <View
            style={
              Object {
                "flexShrink": 1,
                "marginHorizontal": 8,
              }
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 18,
                  "fontWeight": "500",
                }
              }
            >
              Mr. 
              Emre
               
              Abacı
            </Text>
            <Text>
              Vatan Cd
               
              7028
              , 
              Hatay
               
              Elazığ
              , 
              34478
               
              Turkey
            </Text>
            <Text>
              July 1, 1947
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
